cmd/oras/push.go

```diff
@@ -120,94 +120,178 @@ func runPush(opts pushOptions) error {
+	committed := &sync.Map{}
+	opts.PreCopy = display.StatusPrinter("Uploading", verbose)
+	opts.OnCopySkipped = func(ctx context.Context, desc ocispec.Descriptor) error {
+		committed.Store(desc.Digest.String(), desc.Annotations[ocispec.AnnotationTitle])
+		return display.PrintStatus(desc, "Exists   ", verbose)
+	}
+	opts.PostCopy = func(ctx context.Context, desc ocispec.Descriptor) error {
+		committed.Store(desc.Digest.String(), desc.Annotations[ocispec.AnnotationTitle])
+		if err := display.PrintSuccessorStatus(ctx, desc, "Skipped  ", store, committed, verbose); err != nil {
+			return err
 		}
-		desc.Annotations = packOpts.ConfigAnnotations
-		packOpts.ConfigDescriptor = &desc
-		packOpts.PackImageManifest = true
+		return display.PrintStatus(desc, "Uploaded ", verbose)
 	}
-	descs, err := loadFiles(ctx, store, annotations, opts.FileRefs, opts.Verbose)
+}
+
+type packFunc func() (ocispec.Descriptor, error)
+type copyFunc func(desc ocispec.Descriptor) error
+
+func pushArtifact(dst *remote.Repository, pack packFunc, packOpts *oras.PackOptions, copy copyFunc, copyOpts *oras.CopyGraphOptions, verbose bool) (ocispec.Descriptor, error) {
+	root, err := pack()
 	if err != nil {
 		return ocispec.Descriptor{}, err
 	}
 
-	// pack artifact
-	manifestDesc, err := oras.Pack(ctx, store, opts.artifactType, descs, packOpts)
+	copyRootAttempted := false
+	preCopy := copyOpts.PreCopy
+	copyOpts.PreCopy = func(ctx context.Context, desc ocispec.Descriptor) error {
+		if content.Equal(root, desc) {
+			// copyRootAttempted helps track whether the returned error is
+			// generated from copying root.
+			copyRootAttempted = true
+		}
+		if preCopy != nil {
+			return preCopy(ctx, desc)
+		}
+		return nil
+	}
+
+	// push
+	if err = copy(root); err == nil {
+		return root, nil
+	}
+
+	if !copyRootAttempted || !isArtifactUnsupported(err) {
+		return ocispec.Descriptor{}, err
+	}
+
+	if err := display.PrintStatus(root, "Fallback ", verbose); err != nil {
+		return ocispec.Descriptor{}, err
+	}
+	dst.SetReferrersCapability(false)
+	packOpts.PackImageManifest = true
+	root, err = pack()
 	if err != nil {
 		return ocispec.Descriptor{}, err
 	}
 
-	if err = store.Tag(ctx, manifestDesc, tagStaged); err != nil {
+	copyOpts.FindSuccessors = func(ctx context.Context, fetcher content.Fetcher, node ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		if content.Equal(node, root) {
+			// skip non-config
+			content, err := content.FetchAll(ctx, fetcher, root)
+			if err != nil {
+				return nil, err
+			}
+			var manifest ocispec.Manifest
+			if err := json.Unmarshal(content, &manifest); err != nil {
+				return nil, err
+			}
+			return []ocispec.Descriptor{manifest.Config}, nil
+		}
+
+		// config has no successors
+		return nil, nil
+	}
+	if err = copy(root); err != nil {
 		return ocispec.Descriptor{}, err
 	}
-	return manifestDesc, nil
+	return root, nil
+}
+
+func isArtifactUnsupported(err error) bool {
```

Comment:
Curious why we chose names that result in double-negatives?  For example *"is it not not-supported"* vs *"is it supported"*.

Reply:
@nima Nice catch! @shizhMSFT How about renaming it to `NoFallbackToOciImage` or something indicates that we should not retry with OCI Image?

The reason is both `isArtifactUnsupported` and `isArtifactSupported` is not accurate, **we don't know if artifact is supported if the other HTTP error happens**.

Reply:
Moving the disscusion to #672